### DISCUSSION TITLE
Invalid option error in initialization of TimeHarp260

### DISCUSSION THREAD

**Discussion by elerac** [PicoQuant/snAPI#18] [Q&A] [answered]
Hello PicoQuant team!

I'm trying to use TimeHarp260 (PICO) from Python  on Windows. I confirmed the TH260 works with a GUI app downloaded from [the official website](https://www.picoquant.com/products/category/tcspc-and-time-tagging-modules/timeharp-260-tcspc-and-mcs-board-with-pcie-interface#custom1). However, when I try to use it from DLL (it is also downloaded from the official website), the initialization failed and caused an invalid option error. I have tried several situations, `Demo_HistogramSimple.py` in this repository, `histomode.py` and `histomode.c` in [PicoQuant/TH260-Demos](https://github.com/PicoQuant/TH260-Demos). Every situation causes **TH260_ERROR_INVALID_OPTION**. 

Could you please provide me with some guidance? 

Below are the minimum code and printed messages. 

- Windows 11
- Python 3.11.8
- TimeHarp260Lib DLL version 3.2

Code: 
```python
from snAPI.Main import *


sn = snAPI(libType=LibType.TH260)

sn.getDevice()
sn.setLogLevel(logLevel=LogLevel.DataFile, onOff=True)

sn.initDevice(measMode=MeasMode.Histogram)

# T2 and T3 also cause the same error
# sn.initDevice(measMode=MeasMode.T2)
# sn.initDevice(measMode=MeasMode.T3)
```

Printed message: 
```
240323_19:55:18.3666110 INF ----------------------------------------------
240323_19:55:18.3732733 INF INIT
240323_19:55:18.3733105 INF snAPI 1.0.3.6581
240323_19:55:18.3733880 INF API Init 2
240323_19:55:18.3734308 INF API loadSettingsIni C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\snAPI\system.ini
240323_19:55:18.3735665 INF Create C:\Data\PicoQuant\Logs\240323.log
240323_19:55:18.3847995 INF Init Harp library TH260
240323_19:55:18.3848729 DEV TH260_GetLibraryVersion: 3.2
240323_19:55:18.3849101 INF clear manipulators
240323_19:55:18.3849943 INF stringifyConfig: ""
240323_19:55:18.3850485 INF API getDeviceConfig 607
240323_19:55:18.3853383 INF API getDevice ""
240323_19:55:18.3855187 DEV TH260_OpenDevice: 0 1047180 true
240323_19:55:18.3855628 DEV getDevice: "" "1047180"
240323_19:55:18.3855958 INF API getDevice "1047180"
240323_19:55:18.3856310 INF stringifyConfig: "1047180"
240323_19:55:18.3856726 INF API getDeviceConfig 612
240323_19:55:18.3859427 INF API initDevice "1047180" 0 0
240323_19:55:18.3859800 DEV initDevice
240323_19:55:24.6189837 ERR TH260lib error in TH260_Initialize: 0 0 (-19): TH260_ERROR_INVALID_OPTION
240323_19:55:24.6192579 DEV TH260_Initialize failed with:
   libver:  3.2  Feb  4 2020
   devindex :  0
   initmode :  0
   errorcode:  -19
   errorline:  1937
   oktostage:  8
   hilevelinit:  OK
   moduleinfo:   1
   hardwrerev:   1
   bestckcode:   1
   registrev:   6
   firmwrev:   202
   serialnum:   1047180
   hw_version:   0121-0422-02.02
   numchannels:   1
   baseresol:   25 ps
   statusreg[1]:   02030001
   statusreg[2]:   00030001


Lib version: 3.2  Feb  4 2020
240323_19:55:24.6234974 INF API exit
240323_19:55:24.6235537 DEV closeDevices
240323_19:55:24.6235981 DEV TH260_CloseDevice: 0 1047180 false
240323_19:55:24.8313222 INF exit API
240323_19:55:24.8316099 INF API exited
240323_19:55:24.8316467 INF EXIT
240323_19:55:24.8316782 INF ----------------------------------------------
```

**Comment by snizzleorg** [comment, marked as answer]
error -19 means: no license for TTTR or other option 

I checked the serial number of your TH260 and indeed you did not purchase the DLL-option which is necessary to use the TimeHarp260 together with custom written software via the dll. Please send an email to info@picoquant.com mentioning the serial number of your device (1047180) and we will send you a quote for the necessary option.

**Reply by elerac** [reply]
Thank you for your prompt response! 
Our team was not anticipating the need for an additional license. We will send an email to get the quote.

**Reply by elerac** [reply]
@snizzleorg I got the license file, and this issue is solved. I close this discussion. Thank you again!

**Comment by snizzleorg** [comment]
@tpoint75 Maybe the error -19 codes could be a bit more verbatim ?
